glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift

```diff
@@ -0,0 +1,213 @@
+    ///
+    /// - returns: `true` if `now` is past the due hour of the day.
+    func isAfterDueTime(_ now: Date, dueHourOfTheDay: Int = Constants.dueHourOfTheDay) -> Bool {
+        return now > Calendar.current.date(
+            bySettingHour: dueHourOfTheDay,
+            minute: 0,
+            second: 0,
+            of: now
+        )!
+    }
+
+    /// Performs startup checks to decide when to schedule the next metrics ping collection.
+    func schedule() {
+        let now = Date()
+
+        // If the version of the app is different from the last time we ran the app,
+        // schedule the metrics ping for immediate collection. We only need to perform
+        // this check at startup (when overduePingAsFirst is true).
+        if isDifferentVersion() {
+            Dispatchers.shared.serialOperationQueue.addOperation {
+                self.collectPingAndReschedule(now)
+            }
+            return
+        }
+
+        let lastSentDate = getLastCollectedDate()
+        logger.debug("The 'metrics' ping was last sent on \(String(describing: lastSentDate))")
+
+        // We expect to cover 3 cases here:
+        //
+        // (1) - the ping was already collected the current calendar day; only schedule
+        //       one for collecting the next calendar day at the due time;
+        // (2) - the ping was NOT collected the current calendar day, and we're later than
+        //       the due time; collect the ping immediately;
+        // (3) - the ping was NOT collected the current calendar day, but we still have
+        //       some time to the due time; schedule for sending the current calendar day.
+
+        let alreadySentToday = lastSentDate != nil && Calendar.current.isDateInToday(lastSentDate!)
+        if alreadySentToday {
+            // The metrics ping was already sent today. Schedule it for the next
+            // calendar day. This addresses (1).
+            logger.info("The 'metrics' ping was already sent today, \(now).")
+            schedulePingCollection(now, sendTheNextCalendarDay: true)
+        } else if isAfterDueTime(now) {
+            logger.info("The 'metrics' ping is scheduled for immediate collection, \(now)")
+            // **IMPORTANT**
+            //
+            // The reason why we're collecting the "metrics" ping in the `Dispatchers`
+            // queue is that we want to make sure no other metric API adds data before
+            // the ping is collected. All the exposed metrics API dispatch calls to the
+            // engines through the `Dispatchers.API` context, so this ensures we are enqueued
+            // before any other recording API call.
+            //
+            // * Do not change `Dispatchers.shared.serialOperationQueue.addOperation` to
+            // `Dispatchers.shared.launchAPI` as this would break startup overdue ping
+            //  collection. * `addOperation` schedules the task for immediate execution on the
```

Comment:
```suggestion
            //  collection. 
            // * `addOperation` schedules the task for immediate execution on the
```

missing newline here.

Reply:
oh wait, was the `*` meant as a highlight of this paragraph? I assumed this should become a list.
Looks weird inline a bit, but I don't care as much.

Reply:
List is fine with me.  I typically only use the asterisk for second level lists.  This comment was cargo-culted from the Android version so adjusting the formatting for line length may have made it look a bit weird.